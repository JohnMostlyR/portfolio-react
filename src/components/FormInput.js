import React, {Component} from 'react';
import PropTypes from 'prop-types';

class FormInput extends Component {
  static propTypes = {
    helperText: PropTypes.string,
    inputType: PropTypes.string,
    label: PropTypes.string.isRequired,
    maxLength: PropTypes.number,
    minLength: PropTypes.number,
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    isTextArea: PropTypes.bool,
    validate: PropTypes.func,
    value: PropTypes.string,
  };

  static defaultProps = {
    inputType: 'text',
    isTextArea: false,
  };

  state = {
    value: this.props.value,
    error: false,
    _hasFocus: false,
    _height: '',
  };

  componentWillReceiveProps(update) {
    this.setState({value: update.value});
  }

  onChange = (ev) => {
    const value = ev.target.value;
    let _height = '';

    if (this.refs.textAreaElement) {
      if (this.refs.textAreaElement.scrollHeight > this.refs.textAreaElement.clientHeight) {
        _height = this.refs.textAreaElement.scrollHeight;
      } else {
        _height = this.refs.textAreaElement.clientHeight;
      }
    }

    this.setState({value, _height});
  };

  onBlur = (ev) => {
    const name = this.props.name;
    const value = ev.target.value;
    const error = this.props.validate ? this.props.validate(value) : false;
    const _hasFocus = false;

    this.setState({value, error, _hasFocus});

    this.props.onChange({name, value, error});
  };

  onFocus = () => {
    this.setState({_hasFocus: true,});
  };

  setFocus = (ev) => {
    if (this.refs.inputElement) {
      this.refs.inputElement.focus();
    } else if (this.refs.textAreaElement) {
      this.refs.textAreaElement.focus();
    }
  };

  render() {
    const renderInput = () => (
      <input
        className="c-portfolio-form__input"
        type={this.props.inputType}
        name={this.props.name}
        onChange={this.onChange}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        value={this.state.value}
        ref='inputElement'
      />
    );

    const renderTextArea = () => (
      <textarea
        style={{height: this.state._height}}
        className="c-portfolio-form__input c-portfolio-form__input--textarea"
        name={this.props.name}
        onBlur={this.onBlur}
        onChange={this.onChange}
        onFocus={this.onFocus}
        value={this.state.value}
        ref='textAreaElement'
      />
    );

    const renderHelperText = () => {
      if (!this.props.helperText) return '';

      let helperText = this.props.helperText;

      if (this.props.minLength || this.props.maxLength) {
        helperText = `${helperText} Should be between ${this.props.minLength} and ${this.props.maxLength} characters. Is now: ${this.state.value.length}`;
      }

      if (this.state._hasFocus || this.state.error || !this.state.value) {
        return helperText;
      } else if (!this.state.error) {
        return 'Thanks!';
      }

      return '';
    };

    return (
      <div className="c-portfolio-form__input-group">
        <label
          className="c-portfolio-form__label"
          htmlFor="name"
          onClick={this.setFocus}
        >
          <span
            className={`c-portfolio-form__label-content${(this.state._hasFocus || this.state.value) ? ' c-portfolio-form__label-content--on-focus' : ''}`}>{this.props.label}<span
            className="c-portfolio-form__label-content--placeholder"><i>{this.props.placeholder}</i></span></span>
        </label>
        {(this.props.isTextArea) ? renderTextArea() : renderInput()}
        <div className="c-portfolio-form__feedback-group">
          <span
            className={`c-portfolio-form__error-message${(this.state._hasFocus || !this.state.error) ? '' : ' is-active'}`}>{`${this.state.error} `}</span>
          <span
            className={'c-portfolio-form__helper-text is-active'}>{renderHelperText()}</span>
        </div>
      </div>
    );
  }
}

export default FormInput;
