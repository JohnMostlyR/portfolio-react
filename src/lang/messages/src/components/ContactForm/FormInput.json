[
  {
    "id": "portfolio.page.contact.form_input.helper_text.range",
    "defaultMessage": "Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
  },
  {
    "id": "portfolio.page.contact.form_input.helper_text.valid",
    "defaultMessage": "Thanks!"
  }
]