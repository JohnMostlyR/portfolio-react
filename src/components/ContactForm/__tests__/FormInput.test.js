import React from 'react';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';

import FormInput, {Input, Textarea} from '../FormInput';

describe('FormInput', () => {
  const props = {
    helperText: 'This is a helper text',
    inputType: 'dummy',
    label: 'This is the label text',
    maxLength: 15,
    name: 'This is the input its name',
    minLength: 1,
    onChange: jest.fn(),
    placeholder: 'This is the placeholder text',
    isTextArea: true,
    validate: jest.fn(),
    value: 'The value',
  };

  it('should render and match the snapshot', () => {
    const wrapper = shallow(<FormInput name={props.name}
                                       label={props.label}
                                       onChange={props.onChange}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should receive props', () => {
    const wrapper = shallow(<FormInput {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have an initial state', () => {
    const wrapper = shallow(<FormInput {...props} />);

    expect(wrapper.state().value).toBe(props.value);
    expect(wrapper.state().error).toBe(false);
    expect(wrapper.state()._hasFocus).toBe(false);
    expect(wrapper.state()._height).toBe('');
  });

  describe('when a field has focus', () => {
    it('should act when an input field is rendered', () => {
      const wrapper = mount(<FormInput {...props} isTextArea={false}/>);
      wrapper.setState({_hasFocus: false});
      const input = wrapper.find(Input);
      input.simulate('focus', {target: {value: true}});
      expect(wrapper.state()._hasFocus).toBe(true);
      expect(toJson(wrapper)).toMatchSnapshot();
      wrapper.unmount();
    });

    it('should act when a textarea is rendered', () => {
      const wrapper = mount(<FormInput {...props} isTextArea={true}/>);
      wrapper.setState({_hasFocus: false});
      const input = wrapper.find(Textarea);
      input.simulate('focus', {target: {value: true}});
      expect(wrapper.state()._hasFocus).toBe(true);
      expect(toJson(wrapper)).toMatchSnapshot();
      wrapper.unmount();
    });
  });

  describe('when a user populates the input field', () => {
    const wrapper = mount(<FormInput {...props} isTextArea={false}/>);
    const value = 'testing123';

    wrapper.setState({_hasFocus: true});

    beforeEach(() => {
      const input = wrapper.find(Input);
      input.simulate('change', {
        target: {
          value: value,
        },
      });
    });

    it('the user input should be stored in the state property `value`', () => {
      expect(wrapper.state().value).toBe(value);
    });

    it('should update the helper text, when available', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not show any errors as long as the field has the focus', () => {
      wrapper.setState({error: 'Error message.'});

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('and then moves away from the field', () => {
    let wrapper;
    let input;

    beforeEach(() => {
      wrapper = mount(<FormInput {...props} isTextArea={false}/>);
      input = wrapper.find(Input);
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('should validate the input', () => {
      const value = 'This is valid';

      input.simulate('blur', {
        target: {
          value: value,
        },
      });

      const invocationArgs = props.validate.mock.calls[0];
      expect(invocationArgs[0]).toBe(value);
    });

    it('should update state', () => {
      wrapper.setState({
        value: 'nothingness',
        error: false,
        _hasFocus: true,
      });

      const value = 'This is valid';

      input.simulate('blur', {
        target: {
          value: value,
        },
      });

      expect(wrapper.state().value).toBe(value);
      expect(wrapper.state().error).toBeFalsy();
      expect(wrapper.state()._hasFocus).toBe(false);
    });

    it('should show a message that the input is correct', () => {
      wrapper.setState({
        value: 'nothingness',
        error: false,
        _hasFocus: true,
      });

      input.simulate('blur', {
        target: {
          value: 'This is valid',
        },
      });

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should show an error message when the input is not correct', () => {
      props.validate.mockReturnValueOnce('Error, Error, Error');
      wrapper.setState({
        _hasFocus: true,
      });

      input.simulate('blur', {
        target: {
          value: '',
        },
      });

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should call `this.props.onChange` function', () => {
      const invocationArgs = props.onChange.mock.calls[0];

      wrapper.setState({
        value: 'nothingness',
        error: false,
        _hasFocus: true,
      });

      input.simulate('blur', {
        target: {
          value: 'This is valid',
        },
      });

      expect(invocationArgs[0]).toMatchObject({
        name: props.name,
        value: 'This is valid',
        error: undefined,
      });
    });
  });

  describe('when a user populates the textarea', () => {
    const value = 'testing123\ntesting\ntesting\n123';
    let wrapper;

    beforeEach(() => {
      wrapper = mount(<FormInput {...props} isTextArea={true} maxLength={300} />);
      wrapper.setState({_hasFocus: true});

      const textArea = wrapper.find(Textarea);

      textArea.simulate('change', {
        target: {
          value: value,
        },
      });
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('the user input should be stored in the state property `value`', () => {
      expect(wrapper.state().value).toBe(value);
    });

    it('should update the helper text, when available', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not show any errors as long as the field has the focus', () => {
      wrapper.setState({error: 'Error message.'});

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('when the input wraps or with multi-lines', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = mount(<FormInput {...props} isTextArea={true} maxLength={300} />);
        wrapper.setState({_hasFocus: true});
      });

      afterEach(() => {
        wrapper.unmount();
      });

      it('TODO: should adjust its height', () => {
      //  TODO: Test should adjust its height
      })
    })
  });
});