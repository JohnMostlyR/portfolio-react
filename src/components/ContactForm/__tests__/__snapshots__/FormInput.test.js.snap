// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput and then moves away from the field should show a message that the input is correct 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={false}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus="This is valid"
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus="This is valid"
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.input
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="dummy"
        value="This is valid"
      >
        <input
          className="sc-bwzfXH dViZZy"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="dummy"
          value="This is valid"
        />
      </styled.input>
      <div>
        <styled.span
          showError={undefined}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            undefined 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <FormattedMessage
              defaultMessage="Thanks!"
              id="portfolio.page.contact.form_input.helper_text.valid"
            >
              <IntlProvider
                defaultLocale="en"
                locale="en"
              >
                <FormattedMessage
                  defaultMessage="Thanks!"
                  id="portfolio.page.contact.form_input.helper_text.valid"
                  values={Object {}}
                >
                  <span>
                    Thanks!
                  </span>
                </FormattedMessage>
              </IntlProvider>
            </FormattedMessage>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput and then moves away from the field should show an error message when the input is not correct 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={false}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus=""
          >
            <span
              className="sc-bZQynM kNIjCD"
            >
              This is the label text
              <styled.span
                hasFocus=""
              >
                <span
                  className="sc-EHOje jzGrCy"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.input
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="dummy"
        value=""
      >
        <input
          className="sc-bwzfXH dViZZy"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="dummy"
          value=""
        />
      </styled.input>
      <div>
        <styled.span
          showError="Error, Error, Error"
        >
          <span
            className="sc-gzVnrw laxHOg"
          >
            Error, Error, Error 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 0,
                    "maxLength": 15,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 0,
                        "maxLength": 15,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 15 characters. Is now: 0
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput should receive props 1`] = `
<styled.div>
  <styled.label
    htmlFor="name"
    onClick={[Function]}
  >
    <styled.span
      hasFocus="The value"
    >
      This is the label text
      <styled.span
        hasFocus="The value"
      >
        <i>
          This is the placeholder text
        </i>
      </styled.span>
    </styled.span>
  </styled.label>
  <styled.textarea
    innerRef={[Function]}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    style={
      Object {
        "height": "",
      }
    }
    value="The value"
  />
  <div>
    <styled.span
      showError={false}
    >
      false 
    </styled.span>
    <styled.span>
      <FormattedMessage
        defaultMessage="Thanks!"
        id="portfolio.page.contact.form_input.helper_text.valid"
      />
    </styled.span>
  </div>
</styled.div>
`;

exports[`FormInput should render and match the snapshot 1`] = `
<styled.div>
  <styled.label
    htmlFor="name"
    onClick={[Function]}
  >
    <styled.span>
      This is the label text
      <styled.span>
        <i />
      </styled.span>
    </styled.span>
  </styled.label>
  <styled.input
    innerRef={[Function]}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <div>
    <styled.span
      showError={false}
    >
      false 
    </styled.span>
    <styled.span />
  </div>
</styled.div>
`;

exports[`FormInput when a field has focus should act when a textarea is rendered 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={true}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.textarea
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "height": "",
          }
        }
        value="The value"
      >
        <textarea
          className="sc-bxivhb bIGQKc"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "height": "",
            }
          }
          value="The value"
        />
      </styled.textarea>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            false 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 9,
                    "maxLength": 15,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 9,
                        "maxLength": 15,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 15 characters. Is now: 9
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput when a field has focus should act when an input field is rendered 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={false}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.input
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="dummy"
        value="The value"
      >
        <input
          className="sc-bwzfXH dViZZy"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="dummy"
          value="The value"
        />
      </styled.input>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            false 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 9,
                    "maxLength": 15,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 9,
                        "maxLength": 15,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 15 characters. Is now: 9
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput when a user populates the input field should not show any errors as long as the field has the focus 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={false}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.input
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="dummy"
        value="testing123"
      >
        <input
          className="sc-bwzfXH dViZZy"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="dummy"
          value="testing123"
        />
      </styled.input>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            Error message. 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 10,
                    "maxLength": 15,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 10,
                        "maxLength": 15,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 15 characters. Is now: 10
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput when a user populates the input field should update the helper text, when available 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={false}
  label="This is the label text"
  maxLength={15}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.input
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="dummy"
        value="testing123"
      >
        <input
          className="sc-bwzfXH dViZZy"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="dummy"
          value="testing123"
        />
      </styled.input>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            false 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 10,
                    "maxLength": 15,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 10,
                        "maxLength": 15,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 15 characters. Is now: 10
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput when a user populates the textarea should not show any errors as long as the field has the focus 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={true}
  label="This is the label text"
  maxLength={300}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.textarea
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "height": 0,
          }
        }
        value="testing123
      testing
      testing
      123"
      >
        <textarea
          className="sc-bxivhb bIGQKc"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "height": 0,
            }
          }
          value="testing123
        testing
        testing
        123"
        />
      </styled.textarea>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            Error message. 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 30,
                    "maxLength": 300,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 30,
                        "maxLength": 300,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 300 characters. Is now: 30
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;

exports[`FormInput when a user populates the textarea should update the helper text, when available 1`] = `
<FormInput
  helperText="This is a helper text"
  inputType="dummy"
  isTextArea={true}
  label="This is the label text"
  maxLength={300}
  minLength={1}
  name="This is the input its name"
  onChange={[Function]}
  placeholder="This is the placeholder text"
  validate={[Function]}
  value="The value"
>
  <styled.div>
    <div
      className="sc-bdVaJa RMGlR"
    >
      <styled.label
        htmlFor="name"
        onClick={[Function]}
      >
        <label
          className="sc-ifAKCX jZZVKT"
          htmlFor="name"
          onClick={[Function]}
        >
          <styled.span
            hasFocus={true}
          >
            <span
              className="sc-bZQynM kZfPom"
            >
              This is the label text
              <styled.span
                hasFocus={true}
              >
                <span
                  className="sc-EHOje jMKoyM"
                >
                  <i>
                    This is the placeholder text
                  </i>
                </span>
              </styled.span>
            </span>
          </styled.span>
        </label>
      </styled.label>
      <styled.textarea
        innerRef={[Function]}
        name="This is the input its name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "height": 0,
          }
        }
        value="testing123
      testing
      testing
      123"
      >
        <textarea
          className="sc-bxivhb bIGQKc"
          name="This is the input its name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "height": 0,
            }
          }
          value="testing123
        testing
        testing
        123"
        />
      </styled.textarea>
      <div>
        <styled.span
          showError={false}
        >
          <span
            className="sc-gzVnrw ghfcQg"
          >
            false 
          </span>
        </styled.span>
        <styled.span>
          <span
            className="sc-htoDjs bVWJue"
          >
            <span>
              This is a helper text
              Â 
              <FormattedMessage
                defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                id="portfolio.page.contact.form_input.helper_text.range"
                values={
                  Object {
                    "count": 30,
                    "maxLength": 300,
                    "minLength": 1,
                  }
                }
              >
                <IntlProvider
                  defaultLocale="en"
                  locale="en"
                >
                  <FormattedMessage
                    defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
                    id="portfolio.page.contact.form_input.helper_text.range"
                    values={
                      Object {
                        "count": 30,
                        "maxLength": 300,
                        "minLength": 1,
                      }
                    }
                  >
                    <span>
                      Should be between 1 and 300 characters. Is now: 30
                    </span>
                  </FormattedMessage>
                </IntlProvider>
              </FormattedMessage>
            </span>
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</FormInput>
`;
